src/pool-bin/libraries/BinPool.sol

```diff
@@ -184,51 +191,64 @@ library BinPool {
         uint16 binStep;
     }
 
+    struct SwapState {
+        uint24 activeId;
+        uint24 protocolFee;
+        uint24 swapFee;
+        bytes32 feeForProtocol;
+    }
+
     function swap(State storage self, SwapParams memory params, uint128 amountIn)
         internal
-        returns (BalanceDelta result, bytes32 feeForProtocol, uint24 activeId, uint24 swapFee)
+        returns (BalanceDelta result, SwapState memory swapState)
     {
-        if (amountIn == 0) revert BinPool__InsufficientAmountIn();
-
-        activeId = self.slot0.activeId;
+        Slot0 memory slot0Cache = self.slot0;
+        swapState.activeId = slot0Cache.activeId;
         bool swapForY = params.swapForY;
+        swapState.protocolFee =
```

Comment:
(gas savings) 
 
instead of storing protocolFee in swapState, simply
```
uint24 protocolFee =
            swapForY ? slot0Cache.protocolFee.getOneForZeroFee() : slot0Cache.protocolFee.getZeroForOneFee();

``

Reply:
It seems we still need to put it in SwapState otherwise we will need another return value (Swap event needs protocolFee) which results stack too deep error 😂


![CleanShot 2024-05-09 at 11 36 53@2x](https://github.com/pancakeswap/pancake-v4-core/assets/137024020/30ca5189-c002-4232-9be6-00a0f4a0372b)